tools/tcpcong.py

```diff
@@ -0,0 +1,559 @@
+    unsigned __int128 saddr;
+    unsigned __int128 daddr;
+    u16 lport;
+    u16 dport;
+} ipv6_flow_key_t;
+
+typedef struct process_key {
+    char comm[TASK_COMM_LEN];
+    u32  tid;
+} process_key_t;
+
+typedef struct ipv4_flow_val {
+    ipv4_flow_key_t ipv4_key;
+    u16  cong_state;
+} ipv4_flow_val_t;
+
+typedef struct ipv6_flow_val {
+    ipv6_flow_key_t ipv6_key;
+    u16  cong_state;
+} ipv6_flow_val_t;
+
+BPF_HASH(start_ipv4, process_key_t, ipv4_flow_val_t);
+BPF_HASH(start_ipv6, process_key_t, ipv6_flow_val_t);
+SOCK_STORE_DEF
+
+typedef struct data_val {
+    DEF_TEXT
+    u64  last_ts;
+    u16  last_cong_stat;
+} data_val_t;
+
+typedef struct cong {
+    u8  cong_stat:5,
+        ca_inited:1,
+        ca_setsockopt:1,
+        ca_dstlocked:1;
+} cong_status_t;
+
+BPF_HASH(ipv4_stat, ipv4_flow_key_t, data_val_t);
+BPF_HASH(ipv6_stat, ipv6_flow_key_t, data_val_t);
+
+HIST_TABLE
+
+static int entry_state_update_func(struct sock *sk)
+{
+    u16 dport = 0, lport = 0;
+    u32 tid = bpf_get_current_pid_tgid();
+    process_key_t key = {0};
+    bpf_get_current_comm(&key.comm, sizeof(key.comm));
+    key.tid = tid;
+
+    u64 family = sk->__sk_common.skc_family;
+    struct inet_connection_sock *icsk = inet_csk(sk);
+    cong_status_t cong_status;
+    bpf_probe_read_kernel(&cong_status, sizeof(cong_status),
+        (void *)((long)&icsk->icsk_retransmits) - 1);
+    if (family == AF_INET) {
+        ipv4_flow_val_t ipv4_val = {0};
+        ipv4_val.ipv4_key.saddr = sk->__sk_common.skc_rcv_saddr;
+        ipv4_val.ipv4_key.daddr = sk->__sk_common.skc_daddr;
+        ipv4_val.ipv4_key.lport = sk->__sk_common.skc_num;
+        dport = sk->__sk_common.skc_dport;
+        dport = ntohs(dport);
+        lport = ipv4_val.ipv4_key.lport;
+        FILTER_LPORT
+        FILTER_DPORT
+        ipv4_val.ipv4_key.dport = dport;
+        ipv4_val.cong_state = cong_status.cong_stat + 1;
+        start_ipv4.update(&key, &ipv4_val);
+    } else if (family == AF_INET6) {
+        ipv6_flow_val_t ipv6_val = {0};
+        bpf_probe_read_kernel(&ipv6_val.ipv6_key.saddr,
+            sizeof(ipv6_val.ipv6_key.saddr),
+            &sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+        bpf_probe_read_kernel(&ipv6_val.ipv6_key.daddr,
+            sizeof(ipv6_val.ipv6_key.daddr),
+            &sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
+        ipv6_val.ipv6_key.lport = sk->__sk_common.skc_num;
+        dport = sk->__sk_common.skc_dport;
+        dport = ntohs(dport);
+        lport = ipv6_val.ipv6_key.lport;
+        FILTER_LPORT
+        FILTER_DPORT
+        ipv6_val.ipv6_key.dport = dport;
+        ipv6_val.cong_state = cong_status.cong_stat + 1;
+        start_ipv6.update(&key, &ipv6_val);
+    }
+    SOCK_STORE_ADD
+    return 0;
+}
+
+static int ret_state_update_func(struct sock *sk)
+{
+    u64 ts, ts1;
+    u16 family, last_cong_state;
+    u16 dport = 0, lport = 0;
+    u32 tid = bpf_get_current_pid_tgid();
+    process_key_t key = {0};
```

Comment:
If the kernel function is really triggered under process context, you can use tid as key, no need to obtain task comm here.

But I guess the function will be triggered under softirq context, I think you can simply use the struct sock itself as key.

Reply:
I played with this tool locally, all tid are zero :)

Reply:
hi @chenhengqi ,
Do you mean all the state update functions are executed in softirq context?  for your suggestion if  kfunc/kretfunc are not available,  we should use kprobe/kretprobe, right? but when using kprobe/kretprobe we should also use process_key to get the parameter struct sock in kretprobe function. so why should we add extra key type(struct sock) for the hash tables?

Reply:
> I think you can simply use the struct sock itself as key.

Sorry for the noise.

Reply:
Some (if not all) of these functions are run under softirq context, you will get tid=0 and comm=swapper.
Use these values as key is not reliable. You may consider using processor id as key.

Reply:
hi ,
yes, some functions are executed in softirq context.  but if the tid is 0, the pid is also 0.  i think tid + comm is enough as the comm is different for these swapper process.

Reply:
Right the swapper comm typically is swapper/<cpu_id>, so we should be fine here.

Reply:
hi @yonghong-song ,
yes,  thanks for your comment!